aries/aries_vcx_ledger/src/ledger/multi_ledger.rs

```diff
@@ -0,0 +1,670 @@
+
+    async fn get_schema(
+        &self,
+        schema_id: &SchemaId,
+        submitter_did: Option<&Did>,
+    ) -> VcxLedgerResult<Schema> {
+        self.inner.get_schema(schema_id, submitter_did).await
+    }
+
+    async fn get_cred_def(
+        &self,
+        cred_def_id: &CredentialDefinitionId,
+        submitter_did: Option<&Did>,
+    ) -> VcxLedgerResult<CredentialDefinition> {
+        self.inner.get_cred_def(cred_def_id, submitter_did).await
+    }
+
+    async fn get_rev_reg_def_json(
+        &self,
+        rev_reg_id: &RevocationRegistryDefinitionId,
+    ) -> VcxLedgerResult<(
+        RevocationRegistryDefinition,
+        Self::RevocationRegistryDefinitionAdditionalMetadata,
+    )> {
+        let (reg, meta) = self.inner.get_rev_reg_def_json(rev_reg_id).await?;
+
+        Ok((reg, serde_json::to_value(meta)?))
+    }
+    async fn get_rev_reg_delta_json(
+        &self,
+        rev_reg_id: &RevocationRegistryDefinitionId,
+        from: Option<u64>,
+        to: Option<u64>,
+    ) -> VcxLedgerResult<(RevocationRegistryDelta, u64)> {
+        self.inner
+            .get_rev_reg_delta_json(rev_reg_id, from, to)
+            .await
+    }
+
+    async fn get_rev_status_list(
+        &self,
+        rev_reg_id: &RevocationRegistryDefinitionId,
+        timestamp: u64,
+        rev_reg_def_meta: Option<&Self::RevocationRegistryDefinitionAdditionalMetadata>,
+    ) -> VcxLedgerResult<(RevocationStatusList, u64)> {
+        let meta = match rev_reg_def_meta {
+            Some(v) => Some(serde_json::from_value(v.to_owned())?),
+            None => None,
+        };
+
+        self.inner
+            .get_rev_status_list(rev_reg_id, timestamp, meta.as_ref())
+            .await
+    }
+    async fn get_rev_reg(
+        &self,
+        rev_reg_id: &RevocationRegistryDefinitionId,
+        timestamp: u64,
+    ) -> VcxLedgerResult<(RevocationRegistry, u64)> {
+        self.inner.get_rev_reg(rev_reg_id, timestamp).await
+    }
+}
+
+impl<T> AnoncredsLedgerReadAdaptorTrait for AnoncredsLedgerReadAdaptor<T>
+where
+    T: AnoncredsLedgerRead + AnoncredsLedgerSupport,
+    T::RevocationRegistryDefinitionAdditionalMetadata:
+        Serialize + for<'de> Deserialize<'de> + Send + Sync,
+{
+}
+
+#[cfg(test)]
+mod unit_tests {
+    use async_trait::async_trait;
+    use mockall::{mock, predicate::eq};
+    use serde_json::json;
+
+    use super::*;
+
+    mock! {
+        #[derive(Debug)]
+        pub Reader {}
+        #[async_trait]
+        impl AnoncredsLedgerRead for Reader {
+            type RevocationRegistryDefinitionAdditionalMetadata = Value;
+
+            fn get_schema<'life0,'life1,'life2,'async_trait>(&'life0 self,schema_id: &'life1 SchemaId,submitter_did:Option< &'life2 Did> ,) ->  ::core::pin::Pin<Box<dyn ::core::future::Future<Output = VcxLedgerResult<Schema> > + ::core::marker::Send+'async_trait> >where 'life0:'async_trait,'life1:'async_trait,'life2:'async_trait,Self:'async_trait;
```

Comment:
sorry for ugliness in this test code! - sometimes `mock!` macro has trouble when combined with the `async_trait` macro, so the easiest fix is to insert the _fully expanded_ `async_trait` contents. 

i.e. if it was written like this:
```rust
#[async_trait]
impl AnoncredsLedgerRead for Reader
```
then it would expand into what you're seeing below.

Reply:
good call; done